service/dynamodb/dynamodbbatchwriter/batchwriter.go

```diff
@@ -0,0 +1,129 @@
+package dynamodbbatchwriter
+
+import (
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
+	"github.com/aws/aws-sdk-go-v2/service/dynamodb/dynamodbiface"
+)
+
+const defaultFlushAmount = 25
+const defaultRequestBufferCap = 50
+
+// BatchWriter wraps a dynamodb client to expose a simple PutItem/DeleteItem
+// API that buffers requests and takes advantage of BatchWriteItem behind
+// the scenes.
+type BatchWriter struct {
+	_ struct{}
+
+	tableName     string
+	client        dynamodbiface.DynamoDBAPI
+	primaryKeys   []string
+	flushAmount   int
+	requestBuffer []dynamodb.WriteRequest
+}
+
+// New creates a new BatchWriter that will write to table `tableName`
+// using `client`.
+//
+// New will return an error if it fails to access the table information with a
+// DescribeTableRequest.
+func New(tableName string, client dynamodbiface.DynamoDBAPI) (*BatchWriter, error) {
+	describeTableReq := client.DescribeTableRequest(&dynamodb.DescribeTableInput{
+		TableName: &tableName,
+	})
+	describeTableOut, err := describeTableReq.Send()
+	if err != nil {
+		return &BatchWriter{}, err
+	}
+	// List of primary keys of the table. We will get them from a
+	// DescribeTable request to DynamoDB.
+	pKeys := []string{}
+	for _, key := range describeTableOut.Table.KeySchema {
+		pKeys = append(pKeys, *key.AttributeName)
+	}
+	requestBuffer := make(
+		[]dynamodb.WriteRequest, 0, defaultRequestBufferCap,
+	)
+	batchWriter := &BatchWriter{
+		tableName:     tableName,
+		client:        client,
+		primaryKeys:   pKeys,
+		flushAmount:   defaultFlushAmount,
+		requestBuffer: requestBuffer,
+	}
+	return batchWriter, nil
+}
+
+// SetFlushAmount changes the size at which the requestBuffer gets `Flush`ed
+// automatically.
+func (b *BatchWriter) SetFlushAmount(flushAmount int) {
+	b.flushAmount = flushAmount
+}
+
+func (b *BatchWriter) flushIfNeeded() {
+	if len(b.requestBuffer) >= b.flushAmount {
+		b.Flush()
+	}
+}
+
+func (b *BatchWriter) addWriteRequest(wr dynamodb.WriteRequest) {
+	b.requestBuffer = append(b.requestBuffer, wr)
+	b.flushIfNeeded()
+}
+
+// PutItem adds a PutRequest operation to the requestBuffer.
+func (b *BatchWriter) PutItem(putRequest *dynamodb.PutRequest) {
```

Comment:
Rather than having `PutItem` and `DeleteItem`, how do you feel about changing this to be like
`func (b *BatchWriter) Add(item dynamodb.WriteRequest) error`

And have a helper function so the end result would be something like `b.Add(batchwriter.WrapPutItem(putRequest))`. This allows for the service to add new things, while not hindering the functionality of the `BatchWriter`.

Reply:
I actually like the interface as `PutItem` and `DeleteItem` more, because it's more intuitive in my opinion. But if you think having `Add` instead would allow us to add more functionality in the future I'm for it.